### ISSUE TITLE
16.1-Compression: Pruning and the Lottery Ticket Hypothesis

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#83]
很多時候我們會需要把模型用在 Resource-constrained 的環境下
舉例來說 跑在智能手錶/Drone上
只有比較少的 Memory /Computing Power

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1ab697e6-2399-4abd-9cc3-56a7f13c49c8)

Q:為什麼不把資料傳到雲端 直接在雲端上做運算再把結果傳回到 Edge Device
A: Latency 的問題 EX:自駕車的 Sensor , Privacy

五個Network Compression 的技術
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/2a027d38-fdfe-4472-b433-6ff75bfde8ed)

__________

# Network Pruning
Network Pruning 顧名思義就是我們要把 Network 裡面的一些參數把它剪掉

俗話說的好 樹大必有枯枝
一個這麼大的 Network 裡面有很多很多的參數
那每一個參數不一定都有在做事
把沒有用的那些參數把它找出來 把它扔掉

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8bdc1d82-cf64-4ecc-8c63-5d17b2c02f34)

人剛出生的時候 腦袋是空空的 神經元跟神經元間沒什麼連結
在六歲的時候會長出非常多的連結
但是隨著年齡漸長 有一些連結就慢慢消失

那其實 Network Pruning 不是太新的概念
早在這個 90 年代Yann Le Cun 就有一篇 Paper 是講 Network Pruning
那篇 Paper 的 Title是 Optimal Brain Damage
他把它剪掉一些 Weight 看成是一種腦損傷

# Network Pruning的步驟

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/7cd25144-f19d-4202-8f9a-481dcf4127e1)

0.訓練Network 

1.計算參數的重要性

也許可以就把Lifelong Learning每個參數的 bi 算出來
就可以知道那個參數重不重要
我們也可以把神經元當作修剪的單位

2.Pruning

但是你做完這個修剪以後
通常你的正確率就會掉一點
3.把剩餘沒有被剪掉的參數再重新做微調讓正確率再回升一點

重複1~3

Q:為什麼我們不一次剪掉大量的參數?
如果你一次剪掉大量的參數
會對 Network 傷害太大->可能會大到你用 Fine-Tune 也沒有辦法復原

修剪的單位可以以參數為單位  也可以以神經元來當作單位
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/ff214a90-f071-45a4-8de3-d8717e454846)

如果我們現在是以參數當作單位
Network它的形狀會是不規則的->不好實作
GPU 加速的時候是把 Network 的運算看成一個矩陣的乘法 -> 用 GPU 加速也不好加速

所以實際上在做 Weight Pruning 的時候
你可能會把那些 Prune 掉的 Weight 直接補零
這樣的問題是你根本就沒有真的把 Network 變小

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/5b01c7ee-0e1a-45f8-bb97-2d81a2712afe)
紫色的這一條線 = Sparsity = 有多少百分比的參數被 Prune 掉了

這個紫色的這條線它的值都很接近 1
代表有接近大概 95% 以上的參數都被 Prune 掉了

這個 Network Pruning 的方法其實是一個非常有效率的方法
往往你可以 Prune 到 95% 以上的參數
那但是你的 Accuracy 只掉 1~2% 而已

照理說丟掉了 95% 的參數只剩下 5% 的參數
這個 Network 變得很小它的運算要很快吧
但實際上你發現根本就沒有加速多少 甚至可以說根本就沒有加速

那加速的程度要大過 1 才有加速
加速程度小於 1 其實是變慢的

__________

以神經元為單位來做 Pruning也許是一個比較有效的方法
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/40c82318-1cd8-480d-be3e-1ec7780abe52)

如果我們用神經元做單位來 Pruning
丟掉一些神經元以後你 Network 的架構仍然是規則的->你用 PyTorch 比較好實作 也比較好用 GPU 來加速
你實作的時候 你只要改那個每一個 Layer Input Output 的 Dimension 就好了

先 Train 一個大的 Network再把它變小
而且小的 Network 跟大的 Network它們的正確率沒有差太多
Q:我們怎麼不直接Train 一個小的 Network 就好了?
A:如果你直接 Train 一個小的 Network你往往沒有辦法得到跟大的 Network 一樣的正確率

### 大樂透假說
訓練 Network是一個看人品的事情
你抽到一組好的 Initial 的參數你就會得到好的結果
抽到一組壞 Initial 的參數就會得到壞的結果
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1f1da310-d257-490d-afbb-5d21f29aba0c)

當我們去訓練這個大的 Network 的時候
我們等於是在訓練同時訓練很多小的 Network
每一個小的 Network不一定可以成功的被訓練出來

所謂成功的訓練出來是說它不一定可以透過 Gradient Descent找到一個好的 Solution

而今天大的 Network 裡面如果包含的小的 Network 越多
就好像是去買樂透的時候包牌包比較多的彩券
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/05d55c05-5d45-4820-ae9e-9ebf836146a9)

但是大樂透假說它不一定是對的
Rethinking The Value Of Network Pruning
它覺得大樂透假說是觀察到的現象
也許只有在某一些特定的情況下才觀察得到

所以到底大樂透假說是真是假
這個未來尚待更多的研究來證實
